File: misc/LC/1957.go
package main

import (
	"fmt"
	"strings"
	"time"

	"aoc/utils"
)

func makeFancyString(s string) string {
	if len(s) < 3 {
		return s
	}
	var sb strings.Builder
	sb.WriteString(s[:2])
	for i := 2; i < len(s); i++ {
		if s[i] != s[i-1] || s[i] != s[i-2] {
			sb.WriteByte(s[i])
		}
	}
	return sb.String()
}


// Why is this so bad? Is it copying the string every time?
func makeFancyStringBAD(s string) string {
	if len(s) < 3 {
		return s
	}

	out := s[:2]
	for i := 2; i < len(s); i++ {
		if s[i] != s[i-1] || s[i] != s[i-2] {
			out += string(s[i])
		}
	}
	return out
}


// a couple test cases
var testCases = []struct {
	input  string
	output string
}{
	{"leeetcode", "leetcode"},
	{"aaabaaaa", "aabaa"},
	{"aab", "aab"},
	{"a", "a"},
	{"ab", "ab"},
}

func main() {
	// the super long string
	lines := strings.Split(utils.ReadFile("misc/LC/long_string.txt"), "\n")
	in := lines[0]
	out := lines[1]
	fmt.Println("Long string length:", len(in))

	// test the function with the normal test cases
	start := time.Now()
	for _, v := range testCases {
		if makeFancyString(v.input) != v.output {
			println("Test failed")
		}
	}
	elapsed := time.Since(start)
	fmt.Printf("'Good way' execution time: %0.5f\n", elapsed.Seconds())

	// the long string
	start = time.Now()
	if makeFancyString(in) != out {
		println("Test failed")
	}
	elapsed = time.Since(start)
	fmt.Printf(
		"'Good way' execution time (Long string): %0.5f\n", elapsed.Seconds())

	// test the function with the normal test cases
	start = time.Now()
	for _, v := range testCases {
		if makeFancyStringBAD(v.input) != v.output {
			println("Test failed")
		}
	}
	elapsed = time.Since(start)
	fmt.Printf("'Bad way' execution time: %0.5f\n", elapsed.Seconds())

	// the long string
	start = time.Now()
	if makeFancyStringBAD(in) != out {
		println("Test failed")
	}
	elapsed = time.Since(start)
	fmt.Printf(
		"'Bad way' execution time (Long string): %0.5f\n", elapsed.Seconds())
	

}

File: misc/LC/1957.py
# LC 1957 - Delete Characters to Make Fancy String
import time


def makeFancyString(s: str) -> str:
  # Simple way without caring about string building
  if len(s) < 3:
    return s
  out_str = s[:2]
  for i in range(2, len(s)):
    if s[i] != s[i-1] or s[i] != s[i-2]:
      out_str += s[i]
  return out_str



def main():
  tests = [
    ["leeetcode", "leetcode"],
    ["aaabaaaa", "aabaa"],
    ["aab", "aab"],
    ["a", "a"],
    ["", ""]
  ]

  with open("misc/LC/long_string.txt", "r") as f:
    ans, expected = f.readlines()
    tests.append([
      ans.strip().strip('"'),
      expected.strip().strip('"')
      ])
  
  
  for i, (s, expected) in enumerate(tests):
    print(f"Test {i}:")
    if len(s) > 20:
      print(f"  Input: {s[:10]}...{s[-10:]}")
    else:
      print(f"  Input: {s}")
    
    start = time.perf_counter()
    result = makeFancyString(s)
    stop = time.perf_counter()
    print(f"  Time: {stop-start:.8f} sec")
    print(f"  Time: {(stop-start)*1000:.8f} ms")
    assert result == expected, f"{i} failed:" # expected {expected}, got {result}"
  

if __name__ == "__main__":
  main()